Thao-Truong.java

```diff
@@ -0,0 +1,77 @@
+// Find lowest common boss in a tree;
+// Input: [1, 3, 3, -1, -1, 4, 2, 8, 3]
+// 0 reports to 1, 1 reports to 3, 2 reports to 3, 3 reports to -1 (highest boss), etc.
+//      -1
+//     /  \
+//    3    4
+//  / | \   \
+// 8  2  1   5
+// |  |  |
+// 7  6  0
+// Assumption: the input array is non-empty. And  and valid inputs.
+// Requirement: No extra storage. No build-in library.
+
+class Main {
+  public static void main(String[] args) {
+    int[] boss = { 1, 3, 3, -1, -1, 4, 2, 8, 3 };
+    System.out.println(lowestCommonBoss(boss, 1, 1));  // 1
+		System.out.println(lowestCommonBoss(boss, 2, 0));  // 3
+    System.out.println(lowestCommonBoss(boss, 6, 0));  // 3
+    System.out.println(lowestCommonBoss(boss, 7, 1));  // 3
+    System.out.println(lowestCommonBoss(boss, 4, 5));  // 4
+    System.out.println(lowestCommonBoss(boss, 6, 5));  // -1
+    System.out.println(lowestCommonBoss(boss, 4, 2));  // -1
+    System.out.println(lowestCommonBoss(boss, 4, 3));  // -1
+    System.out.println(lowestCommonBoss(boss, 4, -1));  // -1
+  }
+
+  //Time Complexity: O(n^2)
```

Comment:
Time complexity: T(H) = T(H-1) + 2*H = T(H-2) + 2(H-1) + 2H = ... = 2(H + H-1 + H-2 + ... + 1) = O(H^2)

So time complexity is O(H^2) = O(logN*logN) - not O(N^2)

H: height - assuming the tree is balance, H = log(N)
T(H-1): recursion call of lowestCommonBoss
2H: This method calls isBoss twice